src/librustc/traits/select.rs

```diff
@@ -957,9 +957,11 @@ impl<'cx, 'gcx, 'tcx> SelectionContext<'cx, 'gcx, 'tcx> {
         }
 
         if self.can_use_global_caches(param_env) {
-            let mut cache = self.tcx().evaluation_cache.hashmap.borrow_mut();
             if let Some(trait_ref) = self.tcx().lift_to_global(&trait_ref) {
-                cache.insert(trait_ref, WithDepNode::new(dep_node, result));
+                // This may overwrite the cache with the same value
+                self.tcx().evaluation_cache
+                          .hashmap.borrow_mut()
+                         .insert(trait_ref, WithDepNode::new(dep_node, result));
```

Comment:
Should this use `insert_same()`?

Reply:
The type inserted doesn't implement `Eq`, so it cannot use `insert_same`.

Reply:
For this one, implementing `Eq` should be easy. Let's do it.

Reply:
Done.